torch/csrc/autograd/python_cpp_function.cpp

```diff
@@ -88,67 +89,17 @@ void THPCppFunction_dealloc(PyObject* self)
   Py_TYPE(self)->tp_free(self);
 }
 
-PyObject* THPCppFunction_register_hook_dict(PyObject* self, PyObject* _var)
-{
-  if (!THPVariable_Check(_var)) {
-    return PyErr_Format(PyExc_TypeError, "_register_hook_dict expected a variable");
-  }
-  auto var = (THPVariable*)_var;
-  auto& fn = *((THPCppFunction*)self)->cdata;
-  fn.pre_hooks.push_back(std::make_shared<PyFunctionPreHook>(
-      var->backward_hooks, var->cdata->output_nr));
-  Py_RETURN_NONE;
-}
-
-PyObject* THPCppFunction_register_hook(PyObject* self, PyObject* hook)
-{
-  auto& fn = *((THPCppFunction*)self)->cdata;
-  return registerFunctionHook(fn, hook);
-}
-
-PyObject* THPCppFunction_next_functions(THPCppFunction* self, PyObject* hook)
-{
-  auto& next_functions = self->cdata->next_functions;
-  auto num_next = next_functions.size();
-  THPObjectPtr py_functions = PyTuple_New(num_next);
-  if (!py_functions) return NULL;
-  for (size_t i = 0; i < num_next; ++i) {
-    auto& c_tuple = next_functions[i];
-    THPObjectPtr tuple = PyTuple_New(2);
-    if (!tuple) return NULL;
-    PyObject *py_fn = functionToPyObject(c_tuple.first);
-    if (!py_fn) return NULL;
-    PyTuple_SET_ITEM(tuple.get(), 0, py_fn);
-    PyObject *py_idx = PyLong_FromLong(c_tuple.second);
-    if (!py_idx) return NULL;
-    PyTuple_SET_ITEM(tuple.get(), 1, py_idx);
-    PyTuple_SET_ITEM(py_functions.get(), i, tuple.release());
-  }
-  return py_functions.release();
-}
-
-
 } // namespace
 
-static struct PyMethodDef THPCppFunction_methods[] = {
-  {(char*)"_register_hook_dict", (PyCFunction)THPCppFunction_register_hook_dict, METH_O, NULL},
-  {(char*)"register_hook", (PyCFunction)THPCppFunction_register_hook, METH_O, NULL},
-  {NULL}
-};
-
-static struct PyGetSetDef THPCppFunction_properties[] = {
-  {(char*)"next_functions", (getter)THPCppFunction_next_functions, NULL, NULL, NULL},
-  {NULL}
-};
-
-PyTypeObject* _initFunctionPyTypeObject(PyTypeObject& type, const char* name)
+PyTypeObject* _initFunctionPyTypeObject(PyTypeObject& type, const char* name,
+  PyGetSetDef* function_properties, PyMethodDef* function_methods)
 {
   type.tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC;
   type.tp_name = name;
   type.tp_basicsize = sizeof(THPCppFunction);
   type.tp_call = THPCppFunction_call;
-  type.tp_methods = THPCppFunction_methods;
-  type.tp_getset = THPCppFunction_properties;
+  type.tp_methods = function_methods ? function_methods : attributes::default_methods;
```

Comment:
Would it be possible here to add the default methods even when function_methods is provided? That would allow to change the common methods without modifying every function attribute and reduce the chances of forgetting one of the default one?
(same for the properties below)